tests/pytorch/distributed/run_fsdp2_model.py

```diff
@@ -148,24 +160,41 @@ def _train(args):
         assert False
 
     # Apply FSDP/HSDP
-    custom_attrs = save_custom_attrs(model)
-    for sub_module in model.modules():
-        if any(
-            isinstance(sub_module, sub_module_to_wrap) for sub_module_to_wrap in sub_modules_to_wrap
-        ):
-            fully_shard(sub_module, mesh=mesh)
-    fully_shard(model, mesh=mesh)
-    restore_custom_attrs(model, custom_attrs)
+    mp_policy = MixedPrecisionPolicy(
+        param_dtype=torch.bfloat16,
+        reduce_dtype=torch.float32,
+        output_dtype=torch.bfloat16,
+        cast_forward_inputs=True,
+    )
+    custom_attrs = _save_custom_attrs(model)
+    if args.layer_type in [te.MultiheadAttention, te.TransformerLayer]:
+        # Composite modules require wrapping submodules bottom-up for the correct parameter grouping
+        sub_modules_to_wrap = [
+            te.Linear,
+            te.LayerNormLinear,
+            te.LayerNormMLP,
+        ]
+        for sub_module in model.modules():
+            if any(
+                isinstance(sub_module, sub_module_to_wrap)
+                for sub_module_to_wrap in sub_modules_to_wrap
+            ):
+                fully_shard(sub_module, mesh=mesh, mp_policy=None if args.fp8_init else mp_policy)
+    fully_shard(model, mesh=mesh, mp_policy=None if args.fp8_init else mp_policy)
+    _restore_custom_attrs(model, custom_attrs)
 
     optimizer = optim.Adam(model.parameters(), lr=1e-3)
 
     for iteration in range(args.iter):
         # Zero the parameter gradients
         optimizer.zero_grad()
-        input_data = torch.randn(args.batch_size, args.input_size).to(device)
-        output = model(input_data)
-        target = torch.randn(args.batch_size, args.output_size).to(device)
-        loss = F.mse_loss(output, target)
+        input_data = torch.randn(inp_shape).to(device)
+        target = torch.randn(out_shape).to(device)
+        with torch.autograd.detect_anomaly():
+            with torch.amp.autocast(enabled=not args.fp8_init, device_type="cuda"):
+                with te.fp8_autocast(enabled=args.fp8_init, fp8_recipe=fp8_recipe):
+                    output = model(input_data)
+                    loss = F.mse_loss(output, target)
         loss.backward()
```

Comment:
```suggestion
        with torch.autograd.detect_anomaly():
            with torch.amp.autocast(enabled=not args.fp8_init, device_type="cuda"):
                with te.fp8_autocast(enabled=args.fp8_init, fp8_recipe=fp8_recipe):
                    output = model(input_data)
                    loss = F.mse_loss(output, target)
            loss.backward()
```

Reply:
Backward inside context manager as in [example](https://docs.pytorch.org/docs/stable/autograd.html#torch.autograd.detect_anomaly) given in the docs.